### PR TITLE
Update to librustzcash version 0.7, take 2

### LINKED ISSUE #65: Enable the `deny(unsafe_op_in_unsafe_fn)` lint everywhere in zcash-light-client-ffi we define `unsafe` functions
From https://github.com/zcash-hackworks/zcash-light-client-ffi/pull/32/files#r1022962018 :

A side-effect of declaring a function `unsafe` is that unsafe operations in its body are no longer required to be in `unsafe` blocks (see [Rust RFC 2585](https://rust-lang.github.io/rfcs/2585-unsafe-block-in-unsafe-fn.html) for a proposal to change this, not yet adopted). Ideally, we should enable the `deny(unsafe_op_in_unsafe_fn)` lint everywhere we define `unsafe` functions, since the style we write FFI functions in is mostly already consistent with that lint.

### PR DESCRIPTION
**Pull request by sellout**
First time I failed to include the ZIP 317 changes in the feature branch.
closes #65

### CODE REVIEW

**Comment by sellout** [comment]
These changes are identical to #63. That PR has approval from @daira, @pacu, and @str4d. So if one of you could approve this with a message that says it’s fine to carry over those approvals, I can merge this quickly.

File: rust/src/lib.rs
Comment (pacu) [inline rust/src/lib.rs:2219]: can this be refactored to 
````
let input_selector  = if  use_zip317_fees {
 // create ZIP-317 selector
 } else {
  // create ZIP-313 selector
  }
  
// single call to 
   shield_transparent_funds(
                &mut update_ops,
                &network,
                LocalTxProver::new(spend_params, output_params),
                &input_selector,
                &usk,
                &taddrs,
                &memo_bytes,
                ANCHOR_OFFSET,
            )
            .map_err(|e| format_err!("Error while shielding transaction: {}", e))
            
````

**Review by pacu — CHANGES_REQUESTED** [review]
can we remove this repeated code?

File: rust/src/lib.rs
Comment (str4d) [reply rust/src/lib.rs:2219]: Sadly no, because the two branches of the conditional have different types. In the equivalent PR on the Android SDK, I squared the circle (and also deduplicated the logic between `createToAddress` and `shieldToAddress`) with this helper: https://github.com/zcash/zcash-android-wallet-sdk/pull/775/files#diff-ea14a0f1195cb3440fbdf0cfba825a92078a0d1ff9655232f182000d360e2976R1069

**Comment by daira** [comment]
https://github.com/zcash-hackworks/zcash-light-client-ffi/pull/67 fixes issues with `unsafe` annotations (and #65).

**Review by daira — APPROVED** [review]
utACK. (Consider including #67, but it's fine to do that separately.)

**Review by daira — APPROVED** [review]
Re-review after merges: utACK d305bb7e87fc19c4e0d033c214fbc156b81196e6

**Review by pacu — APPROVED** [review]
Tested autoshielding confirmations seem to be fine.

**Pull request merged**
